src/transformers/models/ict/modeling_ict.py

```diff
@@ -0,0 +1,908 @@
+
+        self.add_module("generator", generator)
+        self.add_module("discriminator", discriminator)
+
+        self.add_module("l1_loss", l1_loss)
+        self.add_module("perceptual_loss", perceptual_loss)
+        self.add_module("style_loss", style_loss)
+        self.add_module("adversarial_loss", adversarial_loss)
+
+        self.gen_optimizer = optim.Adam(
+            params=generator.parameters(), lr=float(config.learning_rate), betas=(config.adam_beta1, config.adam_beta2)
+        )
+
+        self.dis_optimizer = optim.Adam(
+            params=discriminator.parameters(),
+            lr=float(config.learning_rate) * float(config.dis_gen_learning_rate),
+            betas=(config.adam_beta1, config.adam_beta2),
+        )
+
+    def process(self, images, edges, masks):
+        self.iteration += 1
+
+        # zero optimizers
+        self.gen_optimizer.zero_grad()
+        self.dis_optimizer.zero_grad()
+
+        # process outputs
+        outputs = self(images, edges, masks)
+        gen_loss = 0
+        dis_loss = 0
+
+        # discriminator loss
+        dis_input_real = images
+        dis_input_fake = outputs.detach()
+        dis_real, _ = self.discriminator(dis_input_real)
+        dis_fake, _ = self.discriminator(dis_input_fake)
+        dis_real_loss = self.adversarial_loss(dis_real, True)
+        dis_fake_loss = self.adversarial_loss(dis_fake, False)
+        dis_loss += (dis_real_loss + dis_fake_loss) / 2
+
+        dis_loss.backward()
+        self.dis_optimizer.step()
+
+        # generator adversarial loss
+        gen_input_fake = outputs
+        gen_fake, _ = self.discriminator(gen_input_fake)
+        gen_gan_loss = self.adversarial_loss(gen_fake, True) * self.config.INPAINT_ADV_LOSS_WEIGHT
+        gen_loss += gen_gan_loss
+
+        # generator l1 loss
+        gen_l1_loss = self.l1_loss(outputs, images) * self.config.L1_LOSS_WEIGHT / torch.mean(masks)
+        gen_loss += gen_l1_loss
+
+        # generator perceptual loss
+        gen_content_loss = self.perceptual_loss(outputs, images)
+        gen_content_loss = gen_content_loss * self.config.CONTENT_LOSS_WEIGHT
+        gen_loss += gen_content_loss
+
+        # generator style loss
+        gen_style_loss = self.style_loss(outputs * masks, images * masks)
+        gen_style_loss = gen_style_loss * self.config.STYLE_LOSS_WEIGHT
+        gen_loss += gen_style_loss
+
+        gen_loss.backward()
+        self.gen_optimizer.step()
+
+        return outputs, gen_loss, dis_loss
+
+    def forward(self, images, edges, masks):
+        images_masked = (images * (1 - masks).float()) + masks
+        inputs = torch.cat((images_masked, edges), dim=1)
+        outputs = self.generator(inputs)
+        return outputs
+
+
+@add_start_docstrings(
+    "The ICTGuidedUpsampler outputting the completed images.",
+    ICT_GUIDED_UP_SAMPLER_START_DOCSTRING,
+)
+class ICTModel(ICTPretrainedGuidedUpsampler):
```

Comment:
This section was modeled after [`Swin2SRForImageSuperResolution`](https://github.com/huggingface/transformers/blob/main/src/transformers/models/swin2sr/modeling_swin2sr.py#L1102-L1215) though I think a new model output type has to be defined...

Reply:
The final model inherits from `XXXPretrainedModel` to initialize the parameters in an optimal way, so this line would be:
`class ICTModel(ICTPretrainedModel):`

You can use the newly added `MaskedImageCompletionOutput` (see this [PR](https://github.com/huggingface/transformers/pull/22152)) as the output class of this model.